support-frontend/assets/helpers/tracking/quantumMetric.ts

```diff
@@ -170,6 +177,57 @@ export function sendEventSubscriptionCheckoutConversion(
 	}
 }
 
+export function sendEventContributionAmountUpdated(
+	amount: number | 'other',
+	contributionType: ContributionType,
+	sourceCurrency: IsoCurrency,
+): void {
+	if (amount === 'other') {
+		return;
+	}
+
+	void canRunQuantumMetric().then((canRun) => {
+		if (canRun) {
+			const sendEventWhenReady = () => {
+				if (window.QuantumMetricAPI?.isOn()) {
+					const sendEventId =
+						contributionType === 'ONE_OFF'
+							? SendEventContributionAmountUpdate.SingleContribution
+							: SendEventContributionAmountUpdate.RecurringContribution;
+					const valueInPence =
+						contributionType === 'ONE_OFF' || contributionType === 'ANNUAL'
+							? amount * 100
+							: amount * 100 * 12;
+					const targetCurrency: IsoCurrency = 'GBP';
+					const convertedValue: number =
+						window.QuantumMetricAPI.currencyConvertFromToValue(
+							valueInPence,
+							sourceCurrency,
+							targetCurrency,
+						);
+					sendEvent(sendEventId, false, convertedValue.toString());
+				}
+			};
+
+			/**
+			 * Quantum Metric's script sets up QuantumMetricAPI.
+			 * We need to check it is defined and ready before we can
+			 * send events to it. If it is ready we call sendEventWhenReady
+			 * immediately. If it is not ready we poll a function that checks
+			 * if QuantumMetricAPI is available. Once it's available we
+			 * call sendEventWhenReady.
+			 */
+			if (window.QuantumMetricAPI?.isOn()) {
+				sendEventWhenReady();
+			} else {
+				waitForQuantumMetricAPi(() => {
+					sendEventWhenReady();
+				});
+			}
+		}
+	});
+}
+
```

Comment:
I wonder if we can reduce a little bit of the duplicated code here with the subs events? I had a go at moving all of the common stuff into the `sendEvent` method.

In that case the send event method becomes:

```ts
function sendEvent(
	id: SendEventId,
	isConversion: boolean,
	value: string,
): void {
	const send = () => {
		if (window.QuantumMetricAPI?.isOn()) {
			window.QuantumMetricAPI.sendEvent(id, isConversion ? 1 : 0, value);
		}
	};

	void canRunQuantumMetric().then((canRun) => {
		if (canRun) {
			/**
			 * Quantum Metric's script sets up QuantumMetricAPI.
			 * We need to check it is defined and ready before we can
			 * send events to it. If it is ready we call sendEventWhenReady
			 * immediately. If it is not ready we poll a function that checks
			 * if QuantumMetricAPI is available. Once it's available we
			 * call sendEventWhenReady.
			 */
			if (window.QuantumMetricAPI?.isOn()) {
				send();
			} else {
				waitForQuantumMetricAPi(() => {
					send();
				});
			}
		}
	});
}
```

In that case, the send subs event then becomes:

```ts
function sendEventSubscriptionCheckoutEvent(
	id: SendEventCheckoutStart | SendEventCheckoutConversion,
	productPrice: ProductPrice,
	billingPeriod: BillingPeriod,
	isConversion: boolean,
): void {
	const sourceCurrency = productPrice.currency;
	const targetCurrency: IsoCurrency = 'GBP';
	const value = getSubscriptionAnnualValue(productPrice, billingPeriod);

	if (!value) {
		return;
	} else if (window.QuantumMetricAPI?.isOn()) {
		const convertedValue: number =
			window.QuantumMetricAPI.currencyConvertFromToValue(
				value,
				sourceCurrency,
				targetCurrency,
			);

		sendEvent(id, isConversion, convertedValue.toString());
	}
}
```

and the contribs event:

```ts
export function sendEventContributionAmountUpdated(
	amount: number | 'other',
	contributionType: ContributionType,
	sourceCurrency: IsoCurrency,
): void {
	if (amount === 'other') {
		return;
	}

	if (window.QuantumMetricAPI?.isOn()) {
		const sendEventId =
			contributionType === 'ONE_OFF'
				? SendEventContributionAmountUpdate.SingleContribution
				: SendEventContributionAmountUpdate.RecurringContribution;
		const valueInPence =
			contributionType === 'ONE_OFF' || contributionType === 'ANNUAL'
				? amount * 100
				: amount * 100 * 12;
		const targetCurrency: IsoCurrency = 'GBP';
		const convertedValue: number =
			window.QuantumMetricAPI.currencyConvertFromToValue(
				valueInPence,
				sourceCurrency,
				targetCurrency,
			);
		sendEvent(sendEventId, false, convertedValue.toString());
	}
}
```

Reply:
I guess this does mean that we interact with the `QuantumMetricAPI` before we check `canRunQuantumMetric`. If we're not allowed to do that, we could always make the `sendEvent` function take a `getEvent` function instead, that we only invoke after the check

Reply:
Yep I spotted the duplicated code and refactored into a function `sendEventWhenReadyTrigger` in this follow-up PR: https://github.com/guardian/support-frontend/pull/3964/files

Reply:
oh nice one!